pymodbus/framer/rtu_framer.py

```diff
@@ -137,31 +129,38 @@ def isFrameReady(self):
 
         :returns: True if ready, False otherwise
         """
-        if len(self._buffer) > self._hsize:
-            if not self._header:
-                self.populateHeader()
+        if len(self._buffer) <= self._hsize:
+            return False
 
-            return self._header and len(self._buffer) >= self._header['len']
-        else:
+        try:
+            # Frame is ready only if populateHeader() successfully populates crc field which finishes RTU frame
+            # Otherwise, if buffer is not yet long enough, populateHeader() raises IndexError
+            self.populateHeader()
+        except IndexError:
             return False
 
+        return True
+
     def populateHeader(self, data=None):
         """
         Try to set the headers `uid`, `len` and `crc`.
 
         This method examines `self._buffer` and writes meta
-        information into `self._header`. It calculates only the
-        values for headers that are not already in the dictionary.
+        information into `self._header`.
 
         Beware that this method will raise an IndexError if
         `self._buffer` is not yet long enough.
         """
-        data = data if data else self._buffer
+        data = data if data is not None else self._buffer
         self._header['uid'] = byte2int(data[0])
         func_code = byte2int(data[1])
         pdu_class = self.decoder.lookupPduClass(func_code)
         size = pdu_class.calculateRtuFrameSize(data)
         self._header['len'] = size
+
+        if len(data) < size:
```

Comment:
Slice operation `data[size - 2:size]` doesn't raise `IndexError` if there is not enough data in array. But other functions in this module expects that this method raises exception in case of incomplete frame. That's why this `if` is added